waspc/cli/src/Wasp/Cli/Command/CreateNewProject/StarterTemplates/StarterTemplateId.hs

```diff
@@ -0,0 +1,115 @@
+{-# OPTIONS_GHC -Wno-unused-top-binds #-}
+
+module Wasp.Cli.Command.CreateNewProject.StarterTemplates.StarterTemplateId
+  ( getStarterTemplateByIdOrThrow,
+    findTemplateByName,
+  )
+where
+
+import Control.Arrow (left)
+import Control.Monad.Except (MonadError, throwError)
+import Data.Foldable (find)
+import Data.Function ((&))
+import Data.List (intercalate, isPrefixOf)
+import Data.Maybe (fromMaybe)
+import StrongPath (Dir', Path', Rel', parseRelDir, reldir)
+import qualified System.FilePath.Posix as FP.Posix
+import qualified Text.Parsec as P
+import Wasp.Cli.Command.CreateNewProject.StarterTemplates.Common (styleText, waspVersionTemplateGitTag)
+import Wasp.Cli.Command.CreateNewProject.StarterTemplates.StarterTemplate (getTemplateName)
+import qualified Wasp.Cli.Command.CreateNewProject.StarterTemplates.StarterTemplate as ST
+import qualified Wasp.Cli.GithubRepo as GhRepo
+
+-- | A way to uniquely reference a specific Wasp starter template.
+-- It is how users specify which starter template they want to use.
+data StarterTemplateId
+  = -- | References one of the featured templates (templates that we know about here in Wasp CLI) by its name.
+    FeaturedTemplateName !String
+  | -- | References any template that is available as a github repo.
+    -- Most useful for referencing third-party repos that we don't know about, since featured ones
+    -- we can more simply reference by name.
+    GhRepoTemplateUri !GhRepo.GithubRepoOwner !GhRepo.GithubRepoName !(Maybe (Path' Rel' Dir'))
+
+-- | This type allows us to reason about types of StarterTemplateIds without having their runtime values.
+data StarterTemplateIdType
+  = IdTypeFeaturedTemplateName
+  | IdTypeGhRepoTemplateUri
+  deriving (Enum, Bounded)
+
+-- | This function serves its purpose just by being defined, even if not used anywhere, because it
+-- ensures (via compiler warning) that we don't forget do update StarterTemplateIdType accordingly
+-- when we change StarterTemplateId (create, delete or modify data constructor).
+getStarterTemplateIdType :: StarterTemplateId -> StarterTemplateIdType
+getStarterTemplateIdType = \case
+  FeaturedTemplateName {} -> IdTypeFeaturedTemplateName
+  GhRepoTemplateUri {} -> IdTypeGhRepoTemplateUri
+
+getStarterTemplateIdTypeDescription :: StarterTemplateIdType -> String
+getStarterTemplateIdTypeDescription = \case
+  IdTypeFeaturedTemplateName -> "a featured template name (e.g. \"saas\")"
+  IdTypeGhRepoTemplateUri -> "a github URI (github:<repo_owner>/<repo_name>[/dir/to/template])"
+
+-- | Given a template id (as string), it will obtain the information on the template that this id references.
+-- It will throw if the id is invalid (can't be parsed, or information on the template can't be obtain based on it).
+getStarterTemplateByIdOrThrow :: (MonadError String m) => [ST.StarterTemplate] -> String -> m ST.StarterTemplate
```

Comment:
Cool way of resolving the Github based templates 👍

Reply:
Thanks! Will likely be throwing away most of this though.